### PR TITLE
Validate blockchain state's ledger hash and staged ledger hash

### PR DESCRIPTION
**Pull request by deepthiskumar**
Ledger hash of the new blockchain state should be the hash of the latest snarked ledger corresponding to a staged ledger.

### CODE REVIEW

File: src/lib/transition_frontier/transition_frontier.ml [hunk 1]
```diff
@@ -304,15 +317,56 @@ struct
...
+                ~default:
+                  (Frozen_ledger_hash.of_ledger_hash
+                     (Ledger.Db.merkle_root t.root_snarked_ledger))
+          | Some proof -> ledger_hash_from_proof proof
+        in
+        if
+          Frozen_ledger_hash.equal target_ledger_hash
+            blockchain_state_ledger_hash
+          && Staged_ledger_hash.equal staged_ledger_hash
+               blockchain_staged_ledger_hash
```
Comment (nholland94) [inline src/lib/transition_frontier/transition_frontier.ml]: Shouldn't we also assert the base of the ledger proof when it's emitted?

File: src/lib/transition_frontier/transition_frontier.ml [hunk 2]
```diff
@@ -304,15 +317,56 @@ struct
...
+        then transitioned_staged_ledger
+        else
+          failwith
+            "Snarked ledger hash and Staged ledger hash after applying the \
+             diff does not match blockchain state's ledger hash and staged \
+             ledger hash resp.\n"
```
Comment (nholland94) [inline src/lib/transition_frontier/transition_frontier.ml]: Actually, one other thing I just realized. This should be a punishment, not a failwith.

File: src/lib/transition_frontier/transition_frontier.ml (hunk 1 above)
Comment (deepthiskumar) [reply src/lib/transition_frontier/transition_frontier.ml]: This is done in `Staged_ledger.apply` as part of checking the scan state invariant

File: src/lib/transition_frontier/transition_frontier.ml (hunk 2 above)
Comment (deepthiskumar) [reply src/lib/transition_frontier/transition_frontier.ml]: I will create an issue for this and make another pr.

**Pull request merged**